File: examples/python/shape.py
from pathlib import Path
import sys

sys.path.insert(0, str(Path(__file__).resolve().parents[2] / "python"))

from stratax import Shape


# Shape stores dimensions only. It does not own array values.
image_shape = Shape([3, 224, 224])
empty_shape = Shape()
zero_sized = Shape([4, 0, 2])
copied = Shape(image_shape)

print("image shape:", image_shape)
print("rank:", image_shape.rank)
print("elements:", image_shape.elements)
print("first dimension:", image_shape[0])
print("last dimension:", image_shape[-1])
print("dimensions:", list(image_shape))
print("copied equals original:", copied == image_shape)

print("empty shape:", empty_shape)
print("zero-sized shape:", zero_sized, "has", zero_sized.elements, "elements")

File: examples/cpp/shape.cpp
#include <iostream>

#include <stratax.h>

int main()
{
    // Shape stores dimensions only. It does not own array values.
    stratax::core::Shape image_shape{3, 224, 224};
    stratax::core::Shape empty_shape{};
    stratax::core::Shape zero_sized{4, 0, 2};
    stratax::core::Shape copied{image_shape};

    std::cout << "image shape: " << image_shape << '\n';
    std::cout << "rank: " << image_shape.rank() << '\n';
    std::cout << "elements: " << image_shape.elements() << '\n';
    std::cout << "first dimension: " << image_shape[0] << '\n';
    std::cout << "last dimension: " << image_shape.at(-1) << '\n';
    std::cout << "dimensions:";
    for (std::size_t dim : image_shape) {
        std::cout << ' ' << dim;
    }
    std::cout << '\n';
    std::cout << "copied equals original: " << (copied == image_shape) << '\n';

    std::cout << "empty shape: " << empty_shape << '\n';
    std::cout << "zero-sized shape: " << zero_sized
              << " has " << zero_sized.elements() << " elements\n";

    return 0;
}
